website/blog/2024/google-search-crawler-draft/index.md

```diff
@@ -0,0 +1,201 @@
+
+### Site Specifics
+
+#### 1. Results are Personalized
+
+Google tries to provide data that's useful to you based on the information it has. This is crucial when working with Google, and you should *always* keep it in mind.
+
+#### 2. Do You Need Personalized Data?
+
+If you want to analyze data that's maximally relevant to your search results, use your main browser's `cookies`.
+
+#### 3. Location Matters
+
+Your IP address's geolocation will influence Google search results.
+
+#### 4. Don't Forget About [`advanced search`](https://www.google.com/advanced_search)
+
+Additional parameters might be very useful for your specific case. For example, if you want to get results predominantly in Spanish, you can use - `https://www.google.com/search?q={YOUR_QUERY}&lr=lang_es`
+
+## 2. Implementing the Crawler
+
+Implemented for [`Crawlee for Python`](https://crawlee.dev/python) v0.4.0
+
+All necessary data can be obtained without browser emulation. We can use `beautifulsoup_crawler` for implementation.
+
+Accordingly, we can use one of Crawlee's basic templates to create our project.
+
+Execute the following commands:
+
+```bash
+pipx install crawlee
+pipx run crawlee create crawlee-google-search
+cd crawlee-google-search
+poetry install
+```
+
+Use whatever tools you're comfortable with - perhaps [`pyenv`](https://github.com/pyenv/pyenv) + ['uv'](https://github.com/astral-sh/uv) or something else.
+
+In an ideal world, when ranking data, we'd want to pass multiple input `QUERIES` and limit the scanning depth with `CRAWL_DEPTH`. Why do we need the fifth page that nobody sees? Or... even the second?
+
+Google may block you if your web scraping is too aggressive, so configure `ConcurrencySettings` for the crawler or use a `proxy`. Using limitations in `ConcurrencySettings` is good practice for any scraping as it helps avoid excessive server load.
+
+Let's implement the necessary code:
+
+```python
+# crawlee-google-search.main
+
+from crawlee.beautifulsoup_crawler import BeautifulSoupCrawler
+from crawlee import Request, ConcurrencySettings
+
+from .routes import router
+
+QUERIES = ["Apify", "Crawlee"]
+
+CRAWL_DEPTH = 2
+
+
+async def main() -> None:
+    """The crawler entry point."""
+    crawler = BeautifulSoupCrawler(
+        request_handler=router,
+        max_request_retries=1,
+        concurrency_settings=ConcurrencySettings(max_concurrency=5, max_tasks_per_minute=20),
+        max_requests_per_crawl=100,
+    )
+
+    requests_lists = [
+        Request.from_url(f"https://www.google.com/search?q={query}",
+                         user_data = {
+                             "current_depth": 0,
+                             "crawl_depth": CRAWL_DEPTH,
+                             "query": query
+                         },
+                         headers = {
+                             "referer": "https://www.google.com/",
```

Comment:
i am missing code explanations and breakdown, we cannot just put the whole code at once and say it a tutorial :D 

valid for the rest of code snippets too.

Reply:
Updated the code snippets.

Also updated the code to better match the latest version of crawlee-python.